### PR TITLE
irmin-bench: run trace from ci

### PR DESCRIPTION
**Pull request by Ngoguey42**
Highlights the problems mentioned in #1428

### CODE REVIEW

File: bench/irmin-pack/trace_replay.ml [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+      && config.empty_blobs = false
+    in
+    Logs.app (fun l ->
+        l "Will %scheck commit hashes against reference."
+          (if check_hash then "" else "NOT "));
+    let commit_seq =
+      open_commit_sequence config.ncommits_trace config.path_conversion
+        config.commit_data_file
+    in
+    let* repo, on_commit, on_end, repo_pp = Store.create_repo ext_config in
```
Comment (Ngoguey42) [inline bench/irmin-pack/trace_replay.ml:359]: only line 346 and 359 changed from tree.exe

File: test/irmin-bench/replay.ml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    let open Fpath in
+    v "test" / "irmin-bench" / "data" / "tezos_actions_1commit.repr"
+    |> to_string
+  in
+  assert (Sys.file_exists trace_path);
+  if Sys.file_exists test_dir then (
+    let cmd = Printf.sprintf "rm -rf %s" test_dir in
+    Logs.debug (fun l -> l "exec: %s\n%!" cmd);
+    let _ = Sys.command cmd in
+    ());
```
Comment (CraigFe) [inline test/irmin-bench/replay.ml:63]:
```suggestion
    ignore (Sys.command cmd : int));
```
Polymorphic ignore is the monster under my bed.

File: test/irmin-bench/replay.ml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    run or is triggered with [dune exec], [dune runtest]. We normalise by
+    switching to the project root first. *)
+let goto_project_root () =
+  let cwd = Fpath.v (Sys.getcwd ()) in
+  match cwd |> Fpath.segs |> List.rev with
+  | "irmin-bench" :: "test" :: "default" :: "_build" :: _ ->
+      let root = cwd |> repeat 4 Fpath.parent in
+      Unix.chdir (Fpath.to_string root)
+  | _ ->
+      ()
```
Comment (CraigFe) [inline test/irmin-bench/replay.ml:49]: Somewhat unrelated, it's possible to generalise this w/ the `__FILE__` macro, so we should maybe have a test-utils private lib some day.

File: bench/irmin-pack/tree.ml [hunk 4]
```diff
@@ -532,14 +142,32 @@ module Bench_suite (Store : Store) = struct
...
+        path_conversion = config.path_conversion;
+        inode_config = config.inode_config;
+        store_type = config.store_type;
+        commit_data_file = config.commit_data_file;
+        artefacts_dir = config.artefacts_dir;
+        keep_store = config.keep_store;
+        keep_stat_trace = config.keep_stat_trace;
+        no_summary = config.no_summary;
+        empty_blobs = config.empty_blobs;
+      }
```
Comment (CraigFe) [inline bench/irmin-pack/tree.ml:159]: This feels off; why do we want both an abstract and a concrete config to convert between?

**Review by CraigFe — COMMENTED** [review]
Thanks for this :-)

File: bench/irmin-pack/tree.ml (hunk 4 above)
Comment (Ngoguey42) [reply bench/irmin-pack/tree.ml:159]: 3 types of config here:
1. `Tree.config` is the mirror of the `tree.exe` CLI parameters
2. `Trace_replay.config` is specific to the trace replay
3. `Store.store_config` is the config to instantiate a repo.

Having both `1.` and `2.` seem unavoidable to me.

Having `3` is annoying, but I didn't find a better way.

The benchmarks are parameterised by the type of store but are also in charge of opening and closing the repo. This means that the configuration for opening the repo is abstract, when seen from the call to `create_repo` (inside `run`), it must be provided to `run`. Hence the config `3.`.

Comment (CraigFe) [reply bench/irmin-pack/tree.ml:159]: Makes sense, thanks. This pattern suggests to me we might want to switch to using objects at some point, but regardless it's not important now.

**Review by CraigFe — APPROVED** [review]

**Comment by CraigFe** [comment]
Rebased to (hopefully) see the CI go green.

**Comment by CraigFe** [comment]
Thanks -- looks like this is validating https://github.com/mirage/irmin/pull/1431 now.

**Pull request merged**